min_heap_binary_tree.rb

```diff
@@ -80,4 +100,95 @@ def parent(index)
   def parents_index(index)
     index / 2
   end
+
+  def left_child_index(index)
+    index * 2
+  end
+
+  def right_child_index(index)
+    (index * 2) + 1
+  end
+
+  def index_to_swap_with_two_children(index)
+    if left_child(index).value >= right_child(index).value
+      (2 * index) + 1
+    else
+      2 * index
+    end
+  end
+
+  def delete_node(delete_index)
+    swap! delete_index, -1
+    tree.pop
+  end
+
+  def parent_child_check(replacement_node, delete_index)
+    byebug
+    if parent(delete_index).value > replacement_node.value
+      swap! delete_index, parents_index(delete_index)
+    elsif has_two_children?(delete_index)
+      replacement_greater_than_children(replacement_node, delete_index)
+    elsif has_only_smaller_left_child?(replacement_node, delete_index)
+      swap! delete_index, left_child_index(delete_index)
+    end
+  end
+
+  def has_only_smaller_left_child?(replacement_node, delete_index)
+    left_child(delete_index) != nil && replacement_node.value > left_child(delete_index).value
+  end
+
+  def only_left_child_smaller?(index)
+    (tree[index].value > left_child(index).value) && (right_child(index).value >= tree[index].value)
+  end
+
+  def only_right_child_smaller?(index)
+    (left_child(index).value > tree[index].value) && (tree[index].value > right_child(index).value)
+  end
+
+  def left_child_larger?(index)
+    left_child(index) != nil && tree[index].value > left_child(index).value
+  end
+
+  def both_children_smaller_swap(replacement_index)
+    index_to_swap = index_to_swap_with_two_children(replacement_index)
+
+    swap! replacement_index, index_to_swap_with_two_children(replacement_index)
+    swap_down(tree[index_to_swap])
+  end
+
+  def two_children_swap(replacement_index)
+    if only_left_child_smaller?(replacement_index)
+      swap! replacement_index, left_child_index(replacement_index)
+    elsif only_right_child_smaller?(replacement_index)
+      swap! replacement_index, right_child_index(replacement_index)
+    else
+      both_children_smaller_swap(replacement_index)
+    end
+  end
+
+  def has_two_children?(replacement_index)
```

Comment:
This method is still getting hit two times, in the same flow, but that is something that does not have to be optimized now.